pkg/statectx/context.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package statectx
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+// ErrCanceled is returned by Context.Err when the context was canceled
+var ErrCanceled = errors.New("job is canceled")
+
+// ErrPaused is returned by Context.Err when the context was paused
+var ErrPaused = errors.New("job is paused")
+
+// statectx.Context implements context.Context interface that acts as context for cancellation signal
+// It also implements:
+// - OnPause()/PauseCtx() that alert in case of pause happened
+// - Any()/AnyCtx() that alert in case of pause/cancel events
+// - State()
+type Context interface {
```

Comment:
i think it's ok but naming is somewhat inconsistent, we have:
```
	case <-ctx.Done():
	case <-ctx.Any():
	case <-ctx.OnPause():
```

i think this would be more consistent and readable:

```
	case <-ctx.Done():
	case <-ctx.Paused(): // instead of OnPause
	case <-ctx.PausedOrDone():  // instead of Any
```